Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107407

>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D48656014](https://our.internmc.facebook.com/intern/diff/D48656014)

eellison committed:
Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor

[ghstack-poisoned]

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -2808,6 +2808,7 @@ class DeviceCachingAllocator {
 
   void release_block(Block* block) {
     TORCH_INTERNAL_ASSERT(!block->expandable_segment_);
+    std::cout << " Freeing tensor \n";
```
Debug print

albanD commented on torch/csrc/cuda/Graph.cpp:
```diff
@@ -30,9 +31,30 @@ void THCPGraph_init(PyObject* module) {
+            } else if (capture_error_mode == "relaxed") {
+              capture_mode = cudaStreamCaptureModeRelaxed;
+            } else {
+              TORCH_INTERNAL_ASSERT(
```
TORCH_CHECK Since this is a user facing API?

albanD commented on torch/csrc/cuda/Graph.cpp:
```diff
@@ -30,9 +31,30 @@ void THCPGraph_init(PyObject* module) {
+            }
+            return self.capture_begin(pool, capture_mode);
+          },
+          py::arg("pool") = (c10::optional<c10::cuda::MempoolId_t>)c10::nullopt,
```
Why was this changed to a nullopt?
Also these args are always passed so no need to provide default for them?

eellison replied:
I was trying to clean up this call: https://github.com/pytorch/pytorch/pull/107407/files#diff-39e542d87359e7d5381d036cbcea9ec759fbe469578bcdc5693ce6cfab7f1a54L72-L77.

albanD replied:
But you still pass None here. So this default value will never be used?
And I guess our custom typecast is already properly translating the `None` that you pass in python into a `nullopt` here.

eellison replied:
Removed arg defaults.

albanD commented on torch/csrc/cuda/Graph.cpp:
```diff
@@ -30,9 +31,30 @@ void THCPGraph_init(PyObject* module) {
+          },
+          py::arg("pool") = (c10::optional<c10::cuda::MempoolId_t>)c10::nullopt,
+          py::arg("capture_error_mode") = "global",
+          py::call_guard<py::gil_scoped_release>())
```
Are you sure this is safe?
This should be quite fast anyways right?

eellison replied:
This was the existing invocation: `torch::wrap_pybind_function_no_gil`

albanD replied:
But you use it around more code here.
If you're sure that absolutely no python code is used here, that's fine!

eellison replied:
Sorry, how do I use it around more code here ? what's the difference ?

albanD replied:
The only difference is that the old code was only releasing for the call to capture_begin.
now the arg parsing and the output conversion (if any) is also covered.
I think the return type is a c++ type so we should be good here.

albanD commented on torch/cuda/graphs.py:
```diff
@@ -68,13 +68,13 @@ def capture_begin(self, pool=None):
+            capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream.
+                Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
+                may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
+                actions in the current thread, and "relaxed" will not error on these actions.
```
Not sure if this is 100% accurate for relaxed? But tbh the nvidia doc is not 100% clear either.
I would just link to the nvidia doc from there and say that our flags match these ones.

Same comment for the doc below

albanD reviewed (commented):
cc @ptrblck can you confirm that this setting is ok to expose and we're not providing a big footgun to users here?

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

albanD commented on torch/cuda/graphs.py:
```diff
@@ -68,13 +68,13 @@ def capture_begin(self, pool=None):
+            capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream.
+                Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
+                may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
+                actions in the current thread, and "relaxed" will not error on these actions. See docs on `cudaStreamCaptureMode`.
```
```suggestion
                actions in the current thread, and "relaxed" will not error on these actions. Do NOT change this setting unless you're very familiar with `cudaStreamCaptureMode <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g9d0535d93a214cbf126835257b16ba85>`_.
```

albanD commented on torch/cuda/graphs.py:
```diff
@@ -139,6 +139,10 @@ class graph:
+        capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream.
+            Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
+            may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
+            actions in the current thread, and "relaxed" will not error on actions. See docs on `cudaStreamCaptureMode`.
```
```suggestion
            actions in the current thread, and "relaxed" will not error on these actions. Do NOT change this setting unless you're very familiar with `cudaStreamCaptureMode <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g9d0535d93a214cbf126835257b16ba85>`_.
```

albanD reviewed (approved):
Small doc suggestion, SGTM otherwise!

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bottler commented:
Why call the option `capture_error_mode` rather than `capture_mode` to match the Nvidia API? If user is launching reliable GPU work on multiple threads, with no errors in any case, this option lets them choose whether to capture only that thread's work or all thread's work, right?

albanD commented:
> If user is launching reliable GPU work on multiple threads, with no errors in any case, this option lets them choose whether to capture only that thread's work or all threads' work, right?

This is NOT the case, and that's why we should add an "Error" in the argument!
The only thing this flag does is to control if unsafe API calls (like cudaMalloc) done in other thread will crash or not.
The capture is always per-stream. And you will always be able to enqueue work on that stream from any thread you want (or only on the current one if you prefer).

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor"


>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.


Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48656014).
